l1/l1.go

```diff
@@ -0,0 +1,70 @@
+package l1
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/NethermindEth/juno/blockchain"
+	"github.com/NethermindEth/juno/core"
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/juno/l1/contract"
+	"github.com/NethermindEth/juno/service"
+	"github.com/NethermindEth/juno/utils"
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	"github.com/ethereum/go-ethereum/event"
+)
+
+//go:generate mockgen -destination=../mocks/mock_subscriber.go -package=mocks github.com/NethermindEth/juno/l1 Subscriber
+type Subscriber interface {
+	WatchLogStateUpdate(opts *bind.WatchOpts, sink chan<- *contract.StarknetLogStateUpdate) (event.Subscription, error)
+}
+
+type Client struct {
+	l1      Subscriber
+	l2Chain *blockchain.Blockchain
+	log     utils.SimpleLogger
+}
+
+var _ service.Service = (*Client)(nil)
+
+func New(l1 Subscriber, chain *blockchain.Blockchain, log utils.SimpleLogger) *Client {
+	return &Client{
+		l1:      l1,
+		l2Chain: chain,
+		log:     log,
+	}
+}
+
+func (c *Client) Run(ctx context.Context) error {
+	subscribeCtx, subscribeCtxCancel := context.WithCancel(ctx)
+	defer subscribeCtxCancel()
+	opts := &bind.WatchOpts{
+		Start:   nil,
+		Context: subscribeCtx,
+	}
+	buffer := 2000
+	updateChan := make(chan *contract.StarknetLogStateUpdate, buffer)
+	sub, err := c.l1.WatchLogStateUpdate(opts, updateChan)
+	if err != nil {
+		return fmt.Errorf("subscribe to L1 state updates: %w", err)
+	}
+	defer sub.Unsubscribe()
+
+	for {
+		select {
+		case err := <-sub.Err():
+			return err
+		case <-ctx.Done():
+			return nil
+		case update := <-updateChan:
```

Comment:
we should check for update.Raw.Removed

https://github.com/ethereum/go-ethereum/blob/master/core/types/log.go#L53-L55